File: node/src/engine/javaRandom.ts
/**
 * Faithful port of `java.util.Random` (48-bit LCG).
 *
 * This guarantees identical `uniform_rand` key sequences across the Java
 * (reference), Ruby, C#, Python and Node engines. Java is the canonical source,
 * so this port reproduces `nextInt(bound)` exactly -- including the 32-bit
 * signed-overflow rejection in the general case, which is what avoids modulo
 * bias. (The Ruby port omits that rejection because Ruby integers are
 * arbitrary-precision; this port emulates the int32 wraparound so it matches
 * the Java reference rather than the Ruby approximation.)
 *
 * BigInt is mandatory for the state, not a stylistic choice: `seed * MULTIPLIER`
 * reaches ~2^83, far past the 2^53 that a JS `number` can represent exactly, so
 * a `number`-based port silently diverges from Java after the first step.
 *
 * @see https://docs.oracle.com/javase/8/docs/api/java/util/Random.html
 */

const MULTIPLIER = 0x5deece66dn;
const ADDEND = 0xbn;
const MASK = (1n << 48n) - 1n;

/** Interpret the low 32 bits of `value` as a signed 32-bit integer. */
export function toInt32(value: number): number {
  return value | 0;
}

export class JavaRandom {
  private seed: bigint;

  constructor(seed: number | bigint) {
    this.seed = JavaRandom.initialScramble(seed);
  }

  setSeed(seed: number | bigint): void {
    this.seed = JavaRandom.initialScramble(seed);
  }

  /** Return a random int in [0, bound) matching Java's nextInt(int). */
  nextInt(bound: number): number {
    if (!Number.isInteger(bound) || bound <= 0) {
      throw new Error(`bound must be a positive integer (got ${bound})`);
    }

    // Power-of-two fast path (matches Java exactly).
    if ((bound & -bound) === bound) {
      return Number((BigInt(bound) * BigInt(this.next(31))) >> 31n);
    }

    // General case: rejection sampling to avoid modulo bias. The rejection
    // condition relies on 32-bit signed overflow, which `| 0` emulates.
    for (;;) {
      const bits = this.next(31);
      const val = bits % bound;
      if (toInt32(bits - val + (bound - 1)) >= 0) return val;
    }
  }

  private static initialScramble(seed: number | bigint): bigint {
    return (BigInt(seed) ^ MULTIPLIER) & MASK;
  }

  /** Java's `protected int next(int bits)`. Exposed for parity tests. */
  next(bits: number): number {
    this.seed = (this.seed * MULTIPLIER + ADDEND) & MASK;
    return Number(this.seed >> BigInt(48 - bits));
  }
}

File: go/internal/engine/javarandom.go
// Package engine contains the workload driving logic: RNG, key generation,
// rate limiting, command selection, and the benchmark orchestrator.
package engine

// JavaRandom is a port of java.util.Random's 48-bit linear congruential
// generator, producing identical sequences to the Java/Ruby/Node/PHP engines.
//
// Go's int64 holds the 48-bit multiply exactly, so — unlike the JavaScript port
// which needs BigInt — plain integer arithmetic with masking is sufficient.
type JavaRandom struct {
	seed int64
}

const (
	jrMultiplier = 0x5DEECE66D
	jrAddend     = 0xB
	jrMask       = (int64(1) << 48) - 1
)

// NewJavaRandom creates a generator seeded like Java's Random(long seed).
func NewJavaRandom(seed int64) *JavaRandom {
	return &JavaRandom{seed: (seed ^ jrMultiplier) & jrMask}
}

// SetSeed re-seeds the generator (matches Java's setSeed).
func (r *JavaRandom) SetSeed(seed int64) {
	r.seed = (seed ^ jrMultiplier) & jrMask
}

// next returns the next `bits` high bits, matching Java's protected next(int).
// The result is returned as a signed 32-bit value (Java's int), so callers can
// reproduce Java's signed arithmetic in nextInt's rejection branch.
func (r *JavaRandom) next(bits uint) int32 {
	r.seed = (r.seed*jrMultiplier + jrAddend) & jrMask
	return int32(r.seed >> (48 - bits))
}

// NextInt returns a uniformly distributed value in [0, bound), matching Java's
// Random.nextInt(int bound), including the modulo-bias rejection loop.
func (r *JavaRandom) NextInt(bound int) int {
	if bound <= 0 {
		panic("bound must be positive")
	}

	// Power-of-two fast path: (bound * next(31)) >> 31.
	if bound&(-bound) == bound {
		return int((int64(bound) * int64(r.next(31))) >> 31)
	}

	// General case with rejection sampling to avoid modulo bias. The rejection
	// test must be done in signed 32-bit arithmetic exactly as Java does it:
	//   while (bits - val + (bound-1) < 0) retry;
	// The subexpression can overflow int32, and that overflow is what triggers a
	// rejection — reproducing it (rather than using wider ints) is required for
	// bit-exact parity.
	for {
		bits := r.next(31)
		val := bits % int32(bound)
		if bits-val+int32(bound-1) >= 0 {
			return int(val)
		}
	}
}
